### PR TITLE
Fix logs sent inside applicationWillTerminate

### PR DESCRIPTION
**Pull request by annakocheshkova**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [X] Are tests passing locally?
* [X] Are the files formatted correctly?
* [x] Did you add unit tests?
* [X] Did you check UI tests on the sample app? They are not executed on CI.
* [X] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Logs sent from `applicationWillTerminate` were not being saved because of `dispatch_async` call. 
To fix this, we add a task to the logsQueue and wait for it inside `applicationWillTerminate` event.

## Related PRs or issues

https://github.com/microsoft/appcenter-sdk-apple/issues/2123/

## Misc

[AB#81803](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/81803)

TODO: unit tests

### CODE REVIEW

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m [hunk 1]
```diff
@@ -177,6 +180,17 @@ - (void)channel:(id<MSChannelProtocol>)channel didResumeWithIdentifyingObject:(i
 
 - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
 
+#if !TARGET_OS_OSX
+  if (isEnabled) {
+    [MS_NOTIFICATION_CENTER addObserver:self
+    selector:@selector(applicationWillTerminate:)
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:186]: Formatting

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m [hunk 2]
```diff
@@ -199,6 +213,15 @@ - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
    */
 }
 
+- (void)applicationWillTerminate:(UIApplication *)application {
+  dispatch_semaphore_wait(self.delayedProcessingSemaphore, dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC));
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:217]: wait should be after `dispatch_async`. This way it won't work

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefaultPrivate.h [hunk 3]
```diff
@@ -18,6 +18,11 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (instancetype)initWithIngestion:(nullable MSAppCenterIngestion *)ingestion;
 
+/**
+* Semaphore for blocking logs queue on applicationWillTerminate.
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefaultPrivate.h:22]: Formatting

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefaultPrivate.h [hunk 4]
```diff
@@ -18,6 +18,11 @@ NS_ASSUME_NONNULL_BEGIN
...
+*/
+@property dispatch_semaphore_t delayedProcessingSemaphore;
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefaultPrivate.h:24]: Wrap into `#if !TARGET_OS_OSX`

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m [hunk 5]
```diff
@@ -177,6 +186,17 @@ - (void)channel:(id<MSChannelProtocol>)channel didResumeWithIdentifyingObject:(i
 
 - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
 
+#if !TARGET_OS_OSX
+  if (isEnabled) {
+    [MS_NOTIFICATION_CENTER addObserver:self
```
Comment (iamclement) [inline AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:191]: We don't really know when the `applicationWillTerminate` of the developer will be called. He can use the delegate but also the notification as we do. Could be before or after. If it's happening after would the log be lost?

Comment (MatkovIvan) [reply AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:191]: yes

Comment (annakocheshkova) [reply AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:191]: Could be. One possible solution is to use a special flag inside our callback and handle it later but seems like a bit of overkill. What do you think?

File: AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m [hunk 6]
```diff
@@ -61,6 +66,26 @@ - (void)tearDown {
   [super tearDown];
 }
 
+#if !TARGET_OS_OSX
+- (void)testAppIsKilled {
+
+  // If
+  [self.sut setEnabled:YES andDeleteDataOnDisabled:YES];
+  id sut = OCMPartialMock(self.sut);
```
Comment (MatkovIvan) [inline AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m:74]: Please explicitly stop mocking at the end

File: AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m [hunk 7]
```diff
@@ -61,6 +66,26 @@ - (void)tearDown {
...
+
+  // When
+  [[NSNotificationCenter defaultCenter] postNotificationName:UIApplicationWillTerminateNotification object:sut];
+
+  // Then
+  OCMVerify([sut applicationWillTerminate:OCMOCK_ANY]);
+  XCTAssertNotNil(self.sut.logsDispatchQueue);
+  [self.sut setEnabled:NO andDeleteDataOnDisabled:YES];
+  [[NSNotificationCenter defaultCenter] postNotificationName:UIApplicationWillTerminateNotification object:sut];
+  OCMReject([sut applicationWillTerminate:OCMOCK_ANY]);
```
Comment (MatkovIvan) [inline AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m:84]: It should be BEFORE usage

Comment (MatkovIvan) [reply AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m:85]: Please add `OCMVerifyAll` as well

File: AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m (hunk 5 above)
Comment (iamclement) [reply AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m:191]: Sounds overkill yes. I wonder if we can simply switch to synchronous storage for all logs when the app enters this phase. The app will be wiped out of the memory so we don't need to think about restoring the async storage. We'll need to make sure previous logs are saved.

If it turns out to introduce too much complexity we'll need to postpone the fix.

**Pull request closed without merging**